cirq-core/cirq/ops/controlled_gate.py

```diff
@@ -177,18 +177,21 @@ def _decompose_with_context_(
                 protocols.unitary(self.sub_gate), control_qubits, qubits[-1]
             )
             return invert_ops + decomposed_ops + invert_ops
-        if isinstance(self.sub_gate, gp.GlobalPhaseGate):
-            # A controlled global phase is a diagonal gate, where each active control value index
-            # is set equal to the phase angle.
+        # Handle global phase gates by checking if the unitary is a 1x1 matrix with a global phase
+        unitary = protocols.unitary(self.sub_gate, default=None)
+        if (
```

Comment:
I was thinking this branch could be merged with the above branch. The conditions `protocols.has_unitary(self.sub_gate)` and `self._qid_shape_() == (2,) * len(self._qid_shape_())` apply to both. The global phase inner branch would then just need an extra check for `protocols.num_qubits(self.sub_gate) == 0`. The `if protocols.is_parameterized(self.sub_gate) or set(shape) != {2}:` check can be removed, because parameterized gates fail `has_unitary`, and `set(shape) == {2}` is equivalent to the `self_qid_shape_()` condition. There's also no need for the three conditions you added, since zero qubits implies a 1x1 unitary, and the abs of a unitary must square-sum to 1 by definition.